### PR TITLE
Adding toroidal harmonic notebooks

### PR DESCRIPTION
**Pull request by clmould**
## Linked Issues

(This merge request is updated version of #3645)

Part of #3340

## Checklist

I confirm that I have completed the following checks:

- [ ] Tests run locally and pass `pytest tests --reactor`
- [ ] Code quality checks run locally and pass `pre-commit run --from-ref develop --to-ref HEAD`
- [ ] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

### CODE REVIEW

**Review by je-cook — COMMENTED** [review]
From my naive understand looks really good. Could have have some tests for the function going into the main package please?

**Review by geograham — CHANGES_REQUESTED** [review]

**Comment by je-cook** [comment]
I was just running our tests on develop and we get a warning from one of the conversion functions (from arccos of a number outside the range `-1<x>1` could we add a np.clip to this line unless that would have unintended side effects?

https://github.com/Fusion-Power-Plant-Framework/bluemira/blob/fcce20f4d6d08157a1902c88ea00b4d48a1655dd/bluemira/utilities/tools.py#L867

to 

```python
np.clip((d_1**2 + d_2**2 - 4 * R_0**2) / (2 * d_1 * d_2), -1, 1)
```
this seem to remove the warning in my quick test.

File: examples/equilibria/single_wire_field_toroidal_harmonics.ex.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+# $P_{\nu}^{\mu}$ is the associated Legendre function of the first kind, of degree $\nu$
+# and order $\mu$. $\textbf{Q}_{\nu}^{\mu}$ is Olver's definition of the associated
+# Legendre function of the second kind. See [here](https://dlmf.nist.gov/14) or F. W. J.
+# Olver
+# (1997b) Asymptotics and Special Functions. A. K. Peters, Wellesley, MA. for more
+# information.
+#
+# The following image shows the psi field we are expecting to obtain at the end of this
+# example.
+
```
Comment (je-cook) [inline examples/equilibria/single_wire_field_toroidal_harmonics.ex.py:82]:
```suggestion
%%
```

File: examples/equilibria/single_wire_field_toroidal_harmonics.ex.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+single_wire_image = Image.open("single_wire_output_image.png")
```
Comment (je-cook) [inline examples/equilibria/single_wire_field_toroidal_harmonics.ex.py:83]: This file doesnt seem to exist

File: bluemira/equilibria/optimisation/harmonics/toroidal_harmonics_approx_functions.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    information.
+
+    """
+    F = 0
+    for s in range(n_max + 1):
+        F += (poch(a, s) * poch(b, s)) / (gamma(c + s) * factorial(s)) * z**s
+    return F
+
+
+def my_legendre_p(lam, mu, x, n_max=20):
```
Comment (je-cook) [inline bluemira/equilibria/optimisation/harmonics/toroidal_harmonics_approx_functions.py:33]: Could we remove the "my_" from the function names now we're no longer testing? Here, any use and the 'my_legrende_q' below
```suggestion
def legendre_p(lam, mu, x, n_max=20):
```

**Review by je-cook — APPROVED** [review]
From a code perspective looks good, thanks for adding the tests. A few little things otherwise from my end good to go. 

I can't comment on the actual maths maybe @CoronelBuendia or @geograham can comment more?

File: bluemira/equilibria/optimisation/harmonics/toroidal_harmonics_approx_functions.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    Parameters
+    ----------
+    lam:
+        degree of the associated Legendre function of the first kind.
+    mu:
+        order -mu of the associated Legendre function of the first kind.
+    x:
+        points at which to evaluate legendreP.
+    n_max:
+        argument required for f_hypergeometric.
```
Comment (geograham) [inline bluemira/equilibria/optimisation/harmonics/toroidal_harmonics_approx_functions.py:75]:
```suggestion
        upper value for summation in f_hypergeometric.
```

File: bluemira/equilibria/optimisation/harmonics/toroidal_harmonics_approx_functions.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    Parameters
+    ----------
+    lam:
+        degree of the associated Legendre function of the second kind.
+    mu:
+        order mu of the associated Legendre function of the second kind.
+    x:
+        points at which to evaluate legendreQ.
+    n_max:
+        argument required for f_hypergeometric.
```
Comment (geograham) [inline bluemira/equilibria/optimisation/harmonics/toroidal_harmonics_approx_functions.py:117]:
```suggestion
        upper value for summation in f_hypergeometric.
```

**Review by geograham — APPROVED** [review]
Looks good to go

**Pull request merged**